### PR TITLE
Add safe succ and pred, deprecated readMay in favor of readMaybe, various minor improvements

### PR DESCRIPTION
**Pull request by minad**

### CODE REVIEW

File: Safe/Foldable.hs [hunk 1]
```diff
@@ -59,15 +62,15 @@ minimumNote, maximumNote :: (Foldable t, Ord a) => String -> t a -> a
 minimumNote note = fromNote note "minimumNote on empty" . minimumMay
 maximumNote note = fromNote note "maximumNote on empty" . maximumMay
 
-minimumByMay, maximumByMay :: (Foldable t) => (a -> a -> Ordering) -> t a -> Maybe a
```
Comment (ndmitchell) [inline Safe/Foldable.hs]: Interesting - my preference is without the brackets, but obviously someone other than me submitting this as a patch, so they got included. I'm happy either way.

**Comment by ndmitchell** [comment]
Thanks for the patch! Some good things in there, and bits I don't agree with so much. In general I'll just list the things I don't like:

I'm generally not a big fan of `-Wall` because some of the changes made to suppress shadowing and unused variables make the code less clear or hide obvious structure by making uninteresting things different - I think that's happened here.

I wouldn't deprecate `readNote` etc because there is some value in them following the same naming convention as the rest of the safe library, and `readDef` and `readNote` still offer different features. I also note my version of `readEither`, namely `read_`, gives much better error messages - so is still better than `readEither`. Certainly a doc comment pointing people at `readEither`/`readMaybe` would be useful even so.

I'm surprised you needed the `sudo` for the Travis build - all my other Travis instances build fine without it...

**Comment by minad** [comment]
Hi, thank you for the review. I knew that some of the changes were controversial - nice to hear your opinion.

* Concerning Wall: Usually I prefer to activate it and just bite the bullet for the "bad warnings" like shadowing... I saw that you disabled the import warning in Safe.Foldable, therefore I thought why not activate it in the cabal? But it is also ok to keep it like that.

* Concerning the read deprecation: I reverted the change. But I have another idea since you mentioned the better error messages - why not expose readEitherSafe as a safe API then so that people using the safe API profit too?

* sudo: It doesn't work for me without it. I thought maybe you activated some magic on your side in some account travis settings? Or old repositories are handled differently since travis changed the sudo behaviour at some point to non-sudo?

**Comment by ndmitchell** [comment]
I always put warnings and Werror in my Travis scripts (it turns on quite a few useful ones) but not in Cabal - that way I see my errors at dev time but my users don't see them at install time.

Exposing some variant of readEither is a good idea. The obvious name is readEither which clashes with base. Is that a problem?

It may well be that old Travis projects had sudo grandfathered in, or it's a user global setting. If that makes it work for more people I'm happy and should add it to all my projects.

**Comment by ndmitchell** [comment]
I should say I won't have access to my dev computer for a few weeks so will have to wait a while to merge/release your improvements.

**Comment by minad** [comment]
Hi,

because of the name clash I named the function readEitherSafe which might not be the best fit. Name clash with base could lead to confusion. I would prefer a different name. What about readEit ;)

No problem if this doesnt get in soon. Have a nice time without dev computer and a happy new year!


Am 31. Dezember 2016 12:51:19 MEZ, schrieb Neil Mitchell <notifications@github.com>:

**Comment by minad** [comment]
@ndmitchell I rebased on current master

**Comment by ndmitchell** [comment]
Thanks a lot - was just about to do that myself :)

**Comment by minad** [comment]
Thx!

**Comment by ndmitchell** [comment]
I appreciate all your revisions. I've released safe-0.3.11 incorporating these changes.

**Comment by minad** [comment]
You're welcome. And thx for the new release :)

**Pull request merged**